exp/api/v1beta1/azuremachinepool_webhook.go

```diff
@@ -257,7 +257,9 @@ func (amp *AzureMachinePool) ValidateOrchestrationMode(c client.Client) func() e
 			}
 			ctx := context.Background()
 			if err := c.Get(ctx, key, ownerMachinePool); err != nil {
```

Comment:
would retrying the get be another option to deal with this? are there terminal situations where the owner isn't found, or are they seemingly transient?

Reply:
My guess is they are transient situations, so perhaps a retry would help. I'll give it a go–it's easy to reproduce here.

Reply:
Maybe @CecileRobertMichon could weigh in from the capz side as to the more CAPI-idiomatic way to deal with a retryable error of this type.

1) return early and requeue a follow-up reconciliation
vs
2) retry in the inner loop of the active reconciliation

Is one of the above definitely more preferred in this scenario?

cc @fabriziopandini @sbueringer just curious about your thoughts

Reply:
I agree retrying would be more resilient than simply ignoring the validation, which might result in invalid configurations being created. This is the webhook so there is no such thing as requeuing. 

Side question: Didn't catch this earlier in the PR that added this, but is the above assuming that the MachinePool name and AzureMachinePool name are the same? If so not sure that's always a valid assumption, there's nothing that prevents a user from naming them differently.

Reply:
@willie-yao had to implement a helper function do find the parent MachinePool for his PR that you might be able to leverage https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2965/files#diff-88cb69b80a9047f0e0cd188951df71c621d843620b72bd52628899b14644209eR100

also the same issue here of the MachinePool not always being found might apply to your scenario @willie-yao FYI

Reply:
Ah, sorry, yes, we're in a webhook, so we're actually enforcing whether or not to initiate the series of events that will result in a potential reconciliation.

The only thing against a retry here is the UX expectation that a webhook would be responsive, so let's not retry for very long.

Reply:
> is the above assuming that the MachinePool name and AzureMachinePool name are the same?

It is. I'll take a look at @willie-yao's helper func.

Yeah since we're in a webhook AFAICT we can only write our own retry logic and do our best to keep it brief.

Reply:
question is: if it fails once, will retrying help? because if objects are created serially then the MachinePool not being present will block the AzureMachinePool from creating which means it will create a deadlock

Reply:
Does this imply that Cluster<---<Provider>Cluster (and other similar CAPI resource<---Provider resource relationships) type validations cannot be implemented via webhooks because we assume that such tightly coupled resource specs are conventionally delivered simultaneously to the apiserver as part of an overall "cluster" template (e.g. multi-resource yaml file)?

Reply:
Retrying does seem to fix things at least in my local testing. Often the parent MP is found on the first try, but sometimes I see this before success:

```
I0125 18:10:52.499411     190 azuremachinepool_webhook.go:258] AzureMachinePoolLogger "msg"="finding parent MachinePool" "attempt"=1
I0125 18:10:53.500315     190 azuremachinepool_webhook.go:258] AzureMachinePoolLogger "msg"="finding parent MachinePool" "attempt"=2
I0125 18:10:54.501515     190 azuremachinepool_webhook.go:258] AzureMachinePoolLogger "msg"="finding parent MachinePool" "attempt"=3
```

Reply:
I still wonder if, after five failed tries, we should return an error or simply log it. Currently we return an error and fail the deployment, but given the loose linkage between the two objects, you could argue that this should be a "best effort" validation.

Reply:
IMO we should return an error, a webhook validation is not meant to be "best effort", especially if we have evidence that a retry does fix things in some cases.

Reply:
I think a short retry is probably fine. I have no idea how your webhook works but you're probably using a cached client. Which essentially means all get/list/requests could get their date from a cache with is slightly stale (usually catches up in <1s)